opal/mca/threads/argobots/threads_argobots_module.c

```diff
@@ -0,0 +1,159 @@
+ * Copyright (c) 2019      Sandia National Laboratories.  All rights reserved.
+ *
+ * Copyright (c) 2020      Cisco Systems, Inc.  All rights reserved.
+ * $COPYRIGHT$
+ *
+ * Additional copyrights may follow
+ *
+ * $HEADER$
+ */
+
+#include "opal_config.h"
+
+#include <unistd.h>
+
+#include "opal/mca/threads/argobots/threads_argobots.h"
+#include "opal/constants.h"
+#include "opal/util/sys_limits.h"
+#include "opal/util/output.h"
+#include "opal/prefetch.h"
+#include "opal/mca/threads/threads.h"
+#include "opal/mca/threads/tsd.h"
+
+#include <abt.h>
+
+struct opal_tsd_key_value {
+    opal_tsd_key_t key;
+    opal_tsd_destructor_t destructor;
+};
+
+static ABT_thread opal_main_thread;
+struct opal_tsd_key_value *opal_tsd_key_values = NULL;
+static int opal_tsd_key_values_count = 0;
+
+/*
+ * Constructor
+ */
+static void opal_thread_construct(opal_thread_t *t)
+{
+    t->t_run = 0;
+    t->t_handle = ABT_THREAD_NULL;
+}
+
+OBJ_CLASS_INSTANCE(opal_thread_t,
+                   opal_object_t,
+                   opal_thread_construct, NULL);
+
+static inline ABT_thread opal_thread_get_argobots_self(void)
+{
+    ABT_thread self;
+    ABT_thread_self(&self);
+    return self;
+}
+
+static void opal_thread_argobots_wrapper(void *arg)
+{
+    opal_thread_t *t = (opal_thread_t *)arg;
+    t->t_ret = ((void *(*)(void *))t->t_run)(t);
+}
+
+opal_thread_t *opal_thread_get_self(void)
+{
+    opal_threads_argobots_ensure_init();
+    opal_thread_t *t = OBJ_NEW(opal_thread_t);
+    t->t_handle = opal_thread_get_argobots_self();
+    return t;
+}
+
+bool opal_thread_self_compare(opal_thread_t *t)
+{
+    opal_threads_argobots_ensure_init();
+    return opal_thread_get_argobots_self() == t->t_handle;
+}
+
+int opal_thread_join(opal_thread_t *t, void **thr_return)
+{
+    opal_threads_argobots_ensure_init();
+    int rc = ABT_thread_free(&t->t_handle);
+    if (thr_return) {
+        *thr_return = t->t_ret;
+    }
+    t->t_handle = ABT_THREAD_NULL;
+    return (0 == rc) ? OPAL_SUCCESS : OPAL_ERROR;
+}
+
+void opal_thread_set_main()
+{
+    opal_threads_argobots_ensure_init();
+    opal_main_thread = opal_thread_get_argobots_self();
+}
+
+int opal_thread_start(opal_thread_t *t)
+{
+    opal_threads_argobots_ensure_init();
+    int rc;
+    if (OPAL_ENABLE_DEBUG) {
+        if (NULL == t->t_run || t->t_handle != ABT_THREAD_NULL) {
+            return OPAL_ERR_BAD_PARAM;
+        }
+    }
+
+    ABT_xstream self_xstream;
+    ABT_xstream_self(&self_xstream);
+    rc = ABT_thread_create_on_xstream(self_xstream,
+                                      opal_thread_argobots_wrapper, t,
+                                      ABT_THREAD_ATTR_NULL, &t->t_handle);
+
+    return (0 == rc) ? OPAL_SUCCESS : OPAL_ERROR;
+}
+
+opal_class_t opal_thread_t_class;
+
+int opal_tsd_key_create(opal_tsd_key_t *key, opal_tsd_destructor_t destructor)
+{
+    opal_threads_argobots_ensure_init();
+    int rc;
+    rc = ABT_key_create(destructor, key);
```

Comment:
This is the same behavior as the present `opal_tsd_key_create()`.
https://github.com/open-mpi/ompi/blob/master/opal/threads/thread.c#L98-L110

This mechanism was introduced by https://github.com/open-mpi/ompi/commit/b3a2bdda7b5457f0335042fcfa89df2b300b92c6#diff-71cc33de6841e88d5ef43fb70cdf9d02.
Regarding the second point, the commit message explains that this TSD registration is needed for only keys created on the main thread in order to suppress memory-leak report.

Reply:
I don't agree with this second part. The underlying problem exists, it simply just not exposed by the current code, because no threads create keys on its own.

In addition to the correctness, my secondary concern is that partially correct code now gets multiplied in all threading components, without a clear explanation.

Reply:
> 1. it is highly unoptimized in the memory management, allocating one key at the time.

I will fix it by doubling the length of `opal_tsd_key_values` when `len(opal_tsd_key_values) < opal_tsd_key_values_count`, which can reduce the number of `realloc()` calls.

> 2. only the main thread register the key in the TSD, while all threads are allowed to create them.
> The underlying problem exists, it simply just not exposed by the current code, because no threads create keys on its own.

I would like to clarify the underlying problem. Please correct me if my understanding is wrong.

- Memory leak.

   - Memory leak of values associated with keys
There is no memory leak concern. In my understanding, management of `opal_tsd_key_values` is basically unnecessary since destructors are called when a thread is freed (no matter whether it is a main thread or not).

   - False-positive memory leak of values associated with keys *when using Valgrind*
This is problematic since Valgrind memory check seems invoked after freeing all the non-main threads but before freeing a main thread.   `opal_tsd_key_values` is maintained only for a main thread in order to suppress this false-positive warning.

   - Memory leak of "keys"
Keys are free by `opal_tsd_key_delete()` independently.

- Timing
If all the TSD destructors need to be called before `MPI_Finalize()`, the current mechanism is broken. I do not think this is the problem, though.

Reply:
I finally understand the underlying issue you mentioned. Please ignore my previous comment.

Yes, in the case where a key is registered by another thread, its destructor should be manually called for the main thread. I don't think creation and deletion of keys are performance critical. I will introduce a critical section.

Reply:
The commit "mca/threads: fix tsd management" addresses the TDS issues.
- All keys are added to `opal_tsd_key_values`, whether the key is created by the main thread or not.
  - Since keys might be registered/freed in parallel, this registration is protected by a lock.
- To reduce the number of `realloc()` calls, the size of `opal_tsd_key_values` is doubled (instead of increasing its size by 1) when `size == count`.